x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,180 @@
+
+package mssql
+
+import (
+	"database/sql"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	s "github.com/elastic/beats/libbeat/common/schema"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	// Register driver.
+	_ "github.com/denisenkom/go-mssqldb"
+)
+
+// NewFetcher is called from all metricsets to initialize their fetching routines. It opens a different connection to
+// the database for each fetch operation
+func NewFetcher(uri string, qs []string, schema *s.Schema, log *logp.Logger) (*Fetcher, error) {
+	db, err := sql.Open("sqlserver", uri)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create db instance")
+	}
+
+	// Check the connection before executing all queries to reduce the number
+	// of connection errors that we might encounter.
+	if err = db.Ping(); err != nil {
+		return nil, err
+	}
+
+	f := &Fetcher{
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	Results     []common.MapStr // Results is the list of metrics returned by the database
+	Error       error           // Error describes all errors that occurred.
+	errs        []error         // errs will accumulate errors returned by each concurrent query in a thread safe manner
+	resultsLock sync.Mutex      // resultsLock protects results and errs for concurrent access while running the queries.
+
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string       // List of queries to execute concurrently.
+	wg      sync.WaitGroup // WaitGroup to wait for all queries to complete.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	// Run queries concurrently.
+	f.doQueries()
+
+	if f.Error != nil {
+		reporter.Error(f.Error)
+		return
+	}
+
+	for _, e := range f.Results {
+		reporter.Event(mb.Event{
+			MetricSetFields: e,
+		})
+	}
+}
+
+// doQueries is executed on object creation from the metricsets via NewFetcher.
+// It will launch a goroutine for each query stored and wait for all results
+// before returning.
+func (f *Fetcher) doQueries() {
+	f.wg.Add(len(f.queries))
+	for _, q := range f.queries {
+		go func() {
```

Comment:
I would suggest for now to not do this in parallel and run one query after the other until we hit a bottle neck. What is the maximum of queries we need to run here? I would expect only 1 or 2.

The reason I would remove it is to have a simpler code.

Reply:
I agree on having a simpler code. To answer your question, the max number of queries I have seen to run really depends on the deep and breath of the module but I'd say around 7-10

Reply:
Concurrency removed